File: grid_walk.cpp
/**
    This is NOT a correct solution! The output from countClimbingStairs overflows if the input n is large enough.
    To solve this, we would need a bigint library. Otherwise, the solution is a correct memoized fibonacci, FWIW.
*/

#include <map>
#include <vector>
#include <set>
#include <iostream>

using namespace std;

std::map<int, unsigned long long int> cache;


int digitsSum(int n) {
    if (n < 0) {
        n = n * -1;
    }
    int sum = 0;
    int d, r;
    while (n > 0) {
        d = n / 10;
        r = n % 10;
        sum += r;
        n -= r;
        if (d > 0) {
            n /= 10;
        }
  }
  return sum;
}

const int LIMIT = 19;

bool accessible(int x, int y) {
    return digitsSum(x) + digitsSum(y) <= LIMIT;
}

void explore(vector<pair<int, int> > &queue, set<pair<int, int> > &visited, int x, int y) {
    pair<int, int> p = make_pair(x, y);
    if (accessible(x, y) && visited.find(p) == visited.end()) {
        visited.insert(p);
        queue.push_back(p);
    }
}

int main(int argc, char *argv[])
{
    vector<pair<int, int> > queue;
    set<pair<int, int> > visited;

    pair<int, int> start = make_pair(0, 0);

    queue.push_back(start);
    visited.insert(start);

    for (int i = 0; i < queue.size(); ++i) {
        int x = queue[i].first;
        int y = queue[i].second;

        explore(queue, visited, x - 1, y);
        explore(queue, visited, x + 1, y);
        explore(queue, visited, x, y - 1);
        explore(queue, visited, x, y + 1);
    }

    std::cout << queue.size() <<  std::endl;
    return 0;
}

File: grid_walk.go
package main

import "fmt"
import "container/list"

type Vertex struct {
	X int
	Y int
}

const LIMIT = 19

func digit_sum(n int) int {
	if n < 0 {
		n = n * -1
	}
	sum := 0
	d, r := 0, 0
	for n > 0 {
		d = n / 10
		r = n % 10
		sum += r
		n -= r
		if d > 0 {
			n /= 10
		}
	}
	return sum
}

func accessible(v Vertex) bool {
	return digit_sum(v.X)+digit_sum(v.Y) <= LIMIT
}

func explore(visited map[Vertex]bool, queue *list.List, v Vertex) {
	_, ok := visited[v]
	if accessible(v) && !ok {
		visited[v] = true
		queue.PushBack(v)
	}
}

func main() {
	start := Vertex{0, 0}
	visited := make(map[Vertex]bool)
	queue := list.New()

	visited[start] = true
	queue.PushBack(start)

	for queue.Len() > 0 {
		next := queue.Front()
		queue.Remove(next)
		nv := next.Value.(Vertex)
		x, y := nv.X, nv.Y

		explore(visited, queue, Vertex{x - 1, y})
		explore(visited, queue, Vertex{x + 1, y})
		explore(visited, queue, Vertex{x, y - 1})
		explore(visited, queue, Vertex{x, y + 1})
	}

	fmt.Println(len(visited))
}
